[WIP] Fix resize approval to work for editing requests

sseago opened the pull request:
The initial "resize approval workflow" feature work is missing some follow-on work needed to allow admin editing of resize/reconfigure requests before approval.

This PR depends on  core PR https://github.com/ManageIQ/manageiq/pull/16381

sseago commented:
FYI, this isn't quite ready to merge. I'm working through some issues that @mansam and I discovered (redirect problems, etc.) I'm working on an update to this.

ZitaNemeckova commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js:
```diff
@@ -1,31 +1,38 @@
+    if (vmCloudResizeFormId == 'new')
+      vm.newRecord = true;
+    else
+      vm.newRecord = false;
```
This can be replaced by `vm.newRecord = vmCloudResizeFormId == 'new';`.

ZitaNemeckova commented:
@miq-bot remove_label pending core

@sseago Please add some steps to reproduce so I can check it does what it's supposed to. Thanks :)

ZitaNemeckova commented:
@miq-bot add_label gaprindashvili/yes

sseago commented:
@ZitaNemeckova To test:
(make sure that there are at least two additional flavors available in addition to the current flavor of your instance)
go to compute -> cloud -> instances
choose instance
configuration -> "reconfigure selected instance"
Choose a flavor to reconfigure to
submit
go to services -> requests
Click on the request. You should be able to edit the request before
approving it. If you make the original reconfigure request as an
admininstrator, the request may auto-approve. If that happens you may
need to try again, starting with a non-admin user.

ZitaNemeckova commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js:
```diff
@@ -1,31 +1,35 @@
-    $http.get('/vm_cloud/resize_form_fields/' + vmCloudResizeFormId)
+    vm.newRecord = vmCloudResizeFormId == 'new';
+
+    $http.get('/vm_cloud/resize_form_fields/' + vmCloudResizeFormId + ',' + vm.objectId)
```
Don't send two ids in one param. Later you have to split them and it's too "hacky". You can send the second id as `'&req_id=' + vm.objectId` as you do in js redirect later. That way it's always clear what it is.

ZitaNemeckova commented on app/controllers/mixins/actions/vm_actions/resize.rb:
```diff
@@ -86,7 +98,9 @@ def resize_vm
           assert_privileges("instance_resize")
-          @record = find_record_with_rbac(VmOrTemplate, params[:id])
+
+          @request_id, record_id = params[:id].split(/\s*,\s*/, 2)
```
Please send those ids in separate params. Do something like this: 
```ruby
record_id = params[:id]
@request_id = params[:req_is]
```
(I hope I didn't switch them in my example :D )

ZitaNemeckova commented:
Before:
<img width="1180" alt="screen shot 2018-03-22 at 9 47 34 am" src="https://user-images.githubusercontent.com/9210860/37761217-5c2811f8-2db9-11e8-86a9-0095acc18a97.png">
After:
<img width="1202" alt="screen shot 2018-03-22 at 10 10 27 am" src="https://user-images.githubusercontent.com/9210860/37761220-619a6dfc-2db9-11e8-87b9-fce6b1e5fcd3.png">
<img width="1185" alt="screen shot 2018-03-22 at 10 10 48 am" src="https://user-images.githubusercontent.com/9210860/37761225-650bbf36-2db9-11e8-9c8b-0ddcdbeb393a.png">
<img width="1185" alt="screen shot 2018-03-22 at 10 12 06 am" src="https://user-images.githubusercontent.com/9210860/37761287-8790b728-2db9-11e8-8610-c2638a06553b.png">

Editing looks good but redirect after submit/cancel is weird. It redirects to instance summary page but it's missing the tree. Do you want it to be redirected there or to previous page (request summary page). If latter you can change this line https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/mixins/actions/vm_actions/resize.rb#L78 to `if @sb[:explorer] && !previous_breadcrumb_url.include?('miq_request')` (which is ugly and any nicer solution is welcome :) ). It would look like this:
![image](https://user-images.githubusercontent.com/9210860/37765556-c060dc58-2dc4-11e8-81f2-ceda2666886d.png)

sseago commented:
So I made the first suggested changes (around the URL param passing). The redirect on submit suggestion didn't work, though -- the issue is that with the suggestion applied, on the initial resize form, that prevents the form from going anywhere. You click submit and the form stays as-is. However irregular the redirect behavior is with the PR as-is, it's better than having a form that doesn't disappear on submit. Also, I remember when first going through this functionality, I was unable to come up with a way that worked as desired for both initial form submission *and* on edit in the approval process.

ZitaNemeckova commented:
@sseago I didn't test my suggestion on initial form (shame on me). I fixed and tested it on both forms. Could you try `if @sb[:explorer] && !(@breadcrumbs.length >= 2 && previous_breadcrumb_url.include?('miq_request'))` instead of`if @sb[:explorer] && !previous_breadcrumb_url.include?('miq_request')`. Thanks :)

sseago commented:
@ZitaNemeckova With the prior update, the suggested changes have been made.

ZitaNemeckova commented:
@sseago Thanks :) I'm not sure about one thing. When editing a request original flavor is not set and it's hard to say which one was set before. Is that an intension or not?

Chosen flavor is unknown:
<img width="1196" alt="screen shot 2018-05-14 at 10 20 45 am" src="https://user-images.githubusercontent.com/9210860/39985987-9cdcbb9c-5760-11e8-82ce-ea1007809899.png">
Chosen flavor is known:
<img width="1196" alt="screen shot 2018-05-14 at 10 21 34 am" src="https://user-images.githubusercontent.com/9210860/39985986-9cc4e4ae-5760-11e8-8c1c-0f712ad96a2d.png">

Otherwise it looks good :)

sseago commented:
@ZitaNemeckova You're right that not showing the original on editing the request is a bug. Digging into the code, it wasn't setting the request_id properly for the angular select element to pick it up.

Unfortunately, after I added the code to track the current value, I still couldn't get it to work right. Everything looks right in the form (checking via debug logs locally, the flavor_id is set on the model, etc.), but it's still not coming through. I've updated the PR to include the part that should set the flavor_id properly, but at least in my dev env, that doesn't seem to be enough to get it to come through. Perhaps I"m missing something obvious here.

ZitaNemeckova commented:
@sseago More like you missed tricky ones :)

`vm.modelCopy` has to be filled after all data are in `vm.CloudModel`
```diff
diff --git a/app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js b/app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js
index d50dc4a..7c17c77 100644
--- a/app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js
+++ b/app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js
@@ -8,7 +8,6 @@ ManageIQ.angular.app.controller('vmCloudResizeFormController', ['$http', '$scope
     vm.flavors = [];
     vm.vmCloudResizeformId = vmCloudResizeFormId;
     vm.objectId = objectId;
-    vm.modelCopy = angular.copy(vm.vmCloudModel);
 
     ManageIQ.angular.scope = vm;
@@ -34,8 +33,9 @@ ManageIQ.angular.app.controller('vmCloudResizeFormController', ['$http', '$scope
 
   function getResizeFormData(response) {
     var data = response.data;
-    vm.flavors = data.flavors
-    vm.vmCloudModel.flavor_id = data.flavor_id
+    vm.flavors = data.flavors;
+    vm.vmCloudModel.flavor_id = data.flavor_id;
+    vm.modelCopy = angular.copy(vm.vmCloudModel);
     miqService.sparkleOff();
   };
```
string vs int 
```diff
diff --git a/app/controllers/mixins/actions/vm_actions/resize.rb b/app/controllers/mixins/actions/vm_actions/resize.rb
index 0ba47dd..7337353 100644
--- a/app/controllers/mixins/actions/vm_actions/resize.rb
+++ b/app/controllers/mixins/actions/vm_actions/resize.rb
@@ -115,7 +115,7 @@ module Mixins
           }
           unless @request_id == 'new'
             @req = MiqRequest.find_by(:id => @request_id)
-            resize_values[:flavor_id] = @req.options[:instance_type]
+            resize_values[:flavor_id] = @req.options[:instance_type].to_i
           end
           render :json => resize_values
         end
 ```
`track by` is kinda problematic. Not sure why but I remember not to use it. `form-changed` is directive that handles form is `pristine` if user sets everything to original value.
```diff
diff --git a/app/views/vm_common/_resize.html.haml b/app/views/vm_common/_resize.html.haml
index 9fa80a7..37a7b3e 100644
--- a/app/views/vm_common/_resize.html.haml
+++ b/app/views/vm_common/_resize.html.haml
@@ -1,6 +1,7 @@
 %form#form_div{"name"          => "angularForm",
                "ng-controller" => "vmCloudResizeFormController as vm",
                "miq-form"      => true,
+               'form-changed'  => '',
                "model"         => "vm.vmCloudModel",
                "model-copy"    => "vm.modelCopy"}
   = render :partial => "layouts/flash_msg"
@@ -18,7 +19,7 @@
       .col-md-8
         %select{:name        => 'flavor_id',
                 'ng-model'   => 'vm.vmCloudModel.flavor_id',
-                'ng-options' => 'flavor.id as flavor.name for flavor in vm.flavors track by flavor.id',
+                'ng-options' => 'flavor.id as flavor.name for flavor in vm.flavors',
                 :miqrequired => true,
                 :checkchange => true}
```

I hope I got them all :)

ZitaNemeckova commented:
LGTM 👍

@sseago Can you please remove wip label? It cannot be merged otherwise.

sseago commented:
@ZitaNemeckova I've removed WIP from the title. I don't think I have permission to remove labels, but perhaps miq-bot removes WIP automatically in response to the title change like it does when adding [WIP]

himdel commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js:
```diff
@@ -1,36 +1,41 @@
     vm.flavors = [];
-    vm.formId = vmCloudResizeFormId;
-    vm.modelCopy = angular.copy(vm.vmCloudModel);
+    vm.vmCloudResizeformId = vmCloudResizeFormId;
```
prefer `formId`, but if you need this name, please make sure this is `vmCloudResizeFormId` and not `...formId`

sseago replied:
I'll fix the capitalization of the 'F' there. With so many moving parts (and a review dragging out six months), I'd rather not mess with the name beyond that and risk breaking something that I'd forgotten about. In the context of this form (just like the reconfigure javascript form controller that I'm not touching), the ID is in fact the request, since that's the way the existing UI is already designed (again, I'm just fixing a bug here, not writing a new feature). It's a "resize instance request" that's being created or edited -- the actual VM doesn't get changed until the request is approved.

himdel replied:
Well, the thing is - there is a possibility you're **introducing** new bugs, just by removing the `formId` as it was before.

So... can you fix the bug without chaning the meaning of `vm.formId`?

himdel commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js:
```diff
@@ -1,36 +1,41 @@
     miqService.sparkleOn();
     var url = '/vm_cloud/resize_vm/' + vmCloudResizeFormId + '?button=submit';
-    miqService.miqAjaxButton(url, vm.vmCloudModel);
+    miqService.miqAjaxButton(url, {objectId: vm.objectId,
```
bad indent

```
miqService.miqAjaxButton(url, {
  objectId: vm.objectId,
  flavor_id: vm.vmCloudModel.flavor_id,
});
```

sseago replied:
fixing...

himdel commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_resize_form_controller.js:
```diff
@@ -1,36 +1,41 @@
     };
     vm.flavors = [];
-    vm.formId = vmCloudResizeFormId;
-    vm.modelCopy = angular.copy(vm.vmCloudModel);
```
missing modelCopy for new

sseago replied:
See comment above from @ZitaNemeckova -- modelCopy has been moved into getResizeFormData since it has to be called after setting everything in vmCloudMOdel

himdel replied:
Ah, sorry about that. Usually `get*FormData` only gets called when editing, not new. 

But you're right, in your case it handles both, withdrawing my objection :).

himdel commented on app/controllers/miq_request_controller.rb:
```diff
@@ -39,7 +39,11 @@ def request_edit
       @_params[:controller] = "vm"
-      reconfigurevms
+      if provision_request.kind_of?(VmCloudReconfigureRequest)
+        resizevms
```
Why this? Is this really the place?

It doesn't sound right that reconfigure also does resize sometimes.

Is this a new feature? (The PR title does not seem to indicate so)

sseago replied:
The back end is "resize" (with request class VmCloudReconfigureRequest) for OpenStack and "reconfigure" (with request class VmReconfigureRequest) for other providers. The issue is that it was decided at some point in the past to expose these two completely different back-end actions as the same thing in the UI. On the model/back end side the actions are completely different, but from a user perspective, they have similar effects (in different providers). The result is that conditionals like this are necessary here.

This is not a new feature -- this PR fixes a bug -- the "edit request" action for administrators does not currently work for resize. A user can request a resize, and an admin can approve it as-is, but without this PR, the request cannot be edited.

himdel replied:
Aah, makes sense, thanks for the explanation! :)

himdel commented on app/controllers/mixins/actions/vm_actions/resize.rb:
```diff
@@ -3,16 +3,26 @@ module Actions
         def resizevms
-          assert_privileges("instance_resize")
-          recs = find_checked_items
+          assert_privileges(params[:pressed])
```
Definitely not!

We can not allow the user to pick which RBAC feature to check for.

This would need a whitelist at least, but .. given the changes, maybe you need a

sseago replied:
@himdel I'm simply copying what's done in reconfigure.rb, which was the primary model for these changes (since both resize and reconfigure use the same UI, exposed as 'reconfigure') -- also, grepping for this, this seems to be a very common pattern. If this isn't the right way to do this, what exactly should be done here? Part of what makes these forms difficult is that the same form has to be used both for the initial request (in the context of the instance/VM ui) as well as by an admin on editing the request (in the context of the requests UI).

himdel replied:
Yes, sorry, it's a pattern we're trying very hard to get rid of.

So... please fix this, by working on this you are in a better position to know the possible privileges and values for `params[:pressed]` than anyone else looking at it later.

Something like..

```
case params[:pressed]
when "your_new_thing"
  assert_privileges("your_new_thing")
else
  assert_privileges("instance_resize")
end
```

would be enough.

sseago replied:
OK. That works for me.

himdel commented on app/controllers/mixins/actions/vm_actions/resize.rb:
```diff
@@ -3,16 +3,26 @@ module Actions
-          recs = find_checked_items
+          assert_privileges(params[:pressed])
+          # if coming in to edit from miq_request list view
+          recs = checked_or_params
```
dead code? `recs` is never used before it is overwritten

himdel replied:
Ah, ok, the difference is that `checked_or_params` will also return `params[:id]` when no checkboxes are found.

Just making sure this was the intended result.

himdel commented on app/controllers/mixins/actions/vm_actions/resize.rb:
```diff
@@ -3,16 +3,26 @@ module Actions
+
           recs = [params[:id].to_i] if recs.blank?
-          @record = find_record_with_rbac(VmOrTemplate, recs.first) # Set the VM object
+          @record ||= find_record_with_rbac(VmOrTemplate, recs.first) # Set the VM object
```
This breaks any code where @record was previously set, are you sure none such exists?

sseago replied:
@record may have been set a few lines above, in which case we want to keep it. This is to handle the case where it was not set. There is a possible edge case if @record is not set above within this method but is already set before -- we could handle that by setting @record to nil before the if block that optionally sets it.

himdel commented on app/controllers/mixins/actions/vm_actions/resize.rb:
```diff
@@ -33,14 +43,18 @@ def resize
             )
           end
           @sb[:explorer] = @explorer
+          @request_id = params[:req_id] ? params[:req_id] : nil
```
this is just `@request_id = params[:req_id]`

sseago replied:
fixing this...

himdel commented on app/views/vm_common/_resize.html.haml:
```diff
@@ -18,7 +19,7 @@
-                'ng-options' => 'flavor.id as flavor.name for flavor in vm.flavors track by flavor.id',
+                'ng-options' => 'flavor.id as flavor.name for flavor in vm.flavors',
                 :miqrequired => true,
                 :checkchange => true}
```
`checkchange` should not be used with `form-changed`, you should be able to remove each checkchange here now :)

sseago replied:
ok...

sseago commented:
@himdel As for the record/request ID thing, having gone through this again, it's become a bit clearer (again, with this review taking six months, some of the details are a bit fuzzy to me right now). The basic summary is this: resize and reconfigure backend actions were required to work within the same front end UI framework. That was the starting requirement for all of this. Reconfigure actually works for both creating and editing requests, and for the existing reconfigure functionality (which I'm not modifying at all) uses the request ID as the ID (since that's the actual object being created and modified). Given all of this, the only straightforward way to get them to work together within the constraints I was working under was to follow the design of the existing working form as much as possible. If I had been free to decouple them and create a separate "resize" action, exposed to the UI as a separate action, this whole thing would have gone differently.

himdel commented:
@sseago I'm sorry ... I getr what you're saying but... correct me if I'm wrong but:

**Statement 1**: `params[:id]` now means something different than before

I think that's true because...

```diff

-  ManageIQ.angular.app.value('vmCloudResizeFormId', '#{@record.id}');
+  ManageIQ.angular.app.value('vmCloudResizeFormId', '#{@request_id || "new"}');
+  ManageIQ.angular.app.value('objectId', '#{@record.id}');
```

and

```diff
-    $http.get('/vm_cloud/resize_form_fields/' + vmCloudResizeFormId)
+    $http.get('/vm_cloud/resize_form_fields/' + vmCloudResizeFormId + '?objectId=' + vm.objectId)
```

**Statement 2**: `reconfigurevms` reads `params[:id]` and was not changed to account for the different meaning

I think that's true just because it does and it wasn't :).


But If both of those are true, then this PR breaks the existing functionality. Am I wrong?

sseago commented:
@himdel I don't believe that it breaks existing functionality because 1) for the 'edit' case, it didn't work before anyway and 2) for the 'new' case, everything that was needed to be changed to reflect the change in meaning of :id should be done already. This change was to bring the partially broken resize UI functionality in line with the fully working reconfigure UI functionality. Do you see anything that looks like it is broken with this change? We've tested the 'new' and 'edit' cases with this PR and both seem to be fully functional.

himdel commented:
> for the 'edit' case, it didn't work before anyway

What didn't work. Resize or reconfigure? If resize, my point still stands.
(If reconfigure, what do you mean by "fully working reconfigure UI functionality"?)

>  Do you see anything that looks like it is broken with this change?

Still not sure what to test actually, the description is not really telling me anything.

himdel commented:
@sseago Maybe what I'm missing could be this..

The screen that's using 'vmCloudResizeFormController' does say "Reconfigure" but maybe this screen is actually resize-only and there is a separate reconfigure screen that does not use any of the changes in your PR? 
If so, would that mean that `resize` and `reconfigure` only really share the call to `request_edit` but are otherwise completely separate? 

If that were so, is is possible that those id changes would indeed not affect reconfigure at all.

mansam commented:
@himdel Reconfigure and Resize use two different views and two different angular controllers. There's some overlap in labeling, but I don't believe they share code.

himdel commented:
Aaah, but I was too rash, sorry :(.

@sseago You promised to fix all the other review comments and didn't.

So, please create a PR to fix:

* keep `vm.formId` in the angular controller, no `vmCloudResizeformId`, especially not with the small `f`
* fix the formatting on line 30 in the controller (https://github.com/ManageIQ/manageiq-ui-classic/pull/2598/files#diff-40f8be8bd21ba536e9696e881ec5b792R30 )
* replace that `assert_privileges(params[:pressed])` with an explicit check for the rbac features we're expecting
* make sure `checked_or_params` giving you `params[:id]` won't break anything, in either version
* fix the problem when @record would already be set
* https://github.com/ManageIQ/manageiq-ui-classic/pull/2598/files#r194740662
* remove `checkchange`

(@simaishi Please don't backport without the second PR.)

mansam commented:
@himdel @sseago went on PTO shortly after you two last exchanged comment, he'll be back to finish the changes next week. Perhaps it would be best for you to revert for now and then merge the completed PR once Scott is back?

himdel commented:
Ah, thanks @mansam .. we'll address the remaining issues in https://github.com/ManageIQ/manageiq-ui-classic/pull/4182 (thanks @ZitaNemeckova ! :))

ZitaNemeckova commented:
> @himdel I don't believe that it breaks existing functionality because 1) for the 'edit' case, it didn't work before anyway and 2) for the 'new' case, everything that was needed to be changed to reflect the change in meaning of :id should be done already. This change was to bring the partially broken resize UI functionality in line with the fully working reconfigure UI functionality. Do you see anything that looks like it is broken with this change? We've tested the 'new' and 'edit' cases with this PR and both seem to be fully functional.

Unfortunately @himdel was right. It no longer works when you want to edit reconfigure request for **Infra** VM. New request for Infra VM reconfigure seems to work.
